### ISSUE TITLE
Problems with dbrestore and single-transaction

### ISSUE THREAD

**Issue by psychok7** [django-dbbackup/django-dbbackup#221]
@ZuluPro So i have been having issues with `dbrestore` (sometimes) and looking deeper i may have found the root problem.

Basically if i did `python3 manage.py dbrestore -i postgres_backup_v13102016_133318.psql` i would run into some errors and my database would not restore. 
The error was `ERROR:  current transaction is aborted, commands ignored until end of transaction block` and `invalid command \N`

Looking deeper i found that this error is being caused by the flag `--single-transaction` . 

To replicate, i took the psql dump (generated by dbbackup) and tried to restore it manually and got the same errors as the db restore command:
`psql --host=smal_postgis --user=postgres --single-transaction postgres < /code/backups/postgres_backup_v13102016_135112.psql`

I then noticed that the `single-transaction` variable was being set to True here https://github.com/django-dbbackup/django-dbbackup/blob/3.0.2/dbbackup/db/postgresql.py#L12

To prove my theory, i ran and it restored fine:
`psql --host=smal_postgis --user=postgres postgres < /code/backups/postgres_backup_v13102016_135112.psql`

Hope this helps fixing the issue or is there a way for me to set `single_transaction` to False with the dbrestore command?

**Comment by psychok7** [comment]
Oh i tried with `django-dbbackup==3.0.2` and `django-dbbackup==3.0.3` and `Django==1.10` and `Postgres==9.5`

**Comment by ZuluPro** [comment]
Hello @psychok7,

Effectively there's a way to unset `single-transaction`:
See http://django-dbbackup.readthedocs.io/en/latest/databases.html#postgresql

I put single transaction by default to avoid inconsistent restore, but maybe it will change.
Otherwise, I'm writing a connector using `pg_restore` and binary backups.

One thing remains, do you know the exact cause of this error ?

**Comment by psychok7** [comment]
@ZuluPro the only errors i see are in the beggining of the restore when the database is empty regarding `indexes`, `relations` or `sequences` like for example `ERROR:  relation "public.users_userprofile" does not exist`.  After the initial errors because it does not exist yet everything runs fine. I don't think there are any errors when there is a database in place.

Maybe make the opposite behavior for `single-transaction` and by default being False??

**Comment by ZuluPro** [comment]
@psychok7 
In fact, I think about change it to `False` but the single transaction behavior is useful for data consistency.

**Comment by psychok7** [comment]
@ZuluPro i did i quick test and my theory is right. 

so basically the problem is with the `does not exist error` when you have no database or no table with that name to drop before starting the restore. for some reason `single-transaction` is not playing right with the drop table. 

do you think there might be a way for single-transaction to ignore the `drop` error in the beggining of the .psql ?? or if in fact there is an error to automatically switch to false??

**Comment by psychok7** [comment]
@ZuluPro by the way with `DBBACKUP_SINGLE_TRANSACTION = False` or `SINGLE_TRANSACTION = False` i get the same problem and my database does not restore

**Comment by ZuluPro** [comment]
@psychok7

 It is a connector settings, in `DBBACKUP_CONNECTORS = {}`

**Comment by psychok7** [comment]
oh ok

**Comment by psychok7** [comment]
@ZuluPro so adding the setting `'SINGLE_TRANSACTION': False` allows me to restore the db with `dbrestore`, although i still see the drop errors in the postgres logs like for example:

```
smal_postgis_1       | ERROR:  extension "postgis_topology" does not exist
smal_postgis_1       | STATEMENT:  DROP EXTENSION postgis_topology;
smal_postgis_1       | ERROR:  extension "postgis" does not exist
smal_postgis_1       | STATEMENT:  DROP EXTENSION postgis;
smal_postgis_1       | ERROR:  schema "tiger" does not exist
smal_postgis_1       | STATEMENT:  DROP SCHEMA tiger;
smal_postgis_1       | ERROR:  schema "public" does not exist
smal_postgis_1       | STATEMENT:  DROP SCHEMA public;
```

Are the dumps being made in a different way then in version below 3.x?

**Comment by ZuluPro** [comment]
@psychok7 

The drop statements are created by `'DROP': True` (default value) of  `DBBACKUP_CONNECTORS`, `--clean` of `pg_dump`.

I think you can avoid this behavior with by set `'DROP': False`, but I'll work to simplify that and make documentation more explicit.

**Comment by psychok7** [comment]
@ZuluPro honestly i prefer having `'DROP': True` as default because i definetly want to delete my old data when i restore my db (and i assume everyone else). I understand the importance of `single-transaction` but if there is no way of making it play nice with the drop command then i prefer not having it.

**Comment by psychok7** [comment]
@ZuluPro hi there, did you address anything related to this on newer releases? I am using  `django-dbbackup==3.0.2` and i just realized DROP is not working for me.

Currently in order to restore as a workaround i have to go into postgres and delete everything first.

my settings are like:
```
DBBACKUP_CONNECTORS = {
    'default': {
        'USER': 'xx',
        'PASSWORD': 'xxx',
        'HOST': 'xx',
        'DROP': True,
        'SINGLE_TRANSACTION': False
    }
}
```

Then i get errors like:

```
smal_postgis_1       | STATEMENT:  CREATE SCHEMA public;
smal_postgis_1       | ERROR:  relation "auth_group" already exists
smal_postgis_1       | STATEMENT:  CREATE TABLE auth_group (
smal_postgis_1       | 	    id integer NOT NULL,
smal_postgis_1       | 	    name character varying(80) NOT NULL
smal_postgis_1       | 	);
```

Any ideas why this is happening??

**Comment by ZuluPro** [comment]
@psychok7 ``DROP``option in PostgreSQL adds ``DROP`` statements inside the dump file.

If you didn't create the dump with this option, it won't work...

**Comment by psychok7** [comment]
@ZuluPro i am creating with the option and i do have DROP statements in the dump.

Fiddling a bit more, i found that the DROP issue might not be because of dbbackup, i manually did a `pg_dump --clean` and the created structure was similar do dbbackup. 

When googling i found some of the errors i am having, for example [here](http://www.postgresqltutorial.com/postgresql-drop-table/) . The thing is the solution seems to be manual and i could not rely on `pg_dump --clean` anymore. (i am not even sure why i have this dependencies issues in the first place)

**Comment by psychok7** [comment]
@ZuluPro to add to the previous comment, the solution that works for me currently is to first drop my database schemas and tables with `drop schema public cascade;` after that i don't run into any errors when restoring (with psql or with dbbackup).

I don't know if this is a common issue or not, or if there is even a better solution but maybe have a setting in dbbackup that would first `drop schema public cascade;` before doing its normal behavior?

**Comment by psychok7** [comment]
@ZuluPro hi, i am testing a new project with `Django 1.11` and this problem still happens. 

Luckily my workaround still works with `django-dbbackup==3.0.2` by setting `'DROP': True` and `'SINGLE_TRANSACTION': False` and deleting manually everything in the database with `drop schema public cascade;` before doing a `dbrestore`.

But with the latest version `django-dbbackup==3.1.3` it does not work. i run into "new" problems (tried both with data in my db, and with a clear db and the restore does not work in any case):

```python


root@644a3e82f4f5:/code# python3 manage.py dbrestore --settings=proj.settings.local
Finding latest backup
Restoring backup for database 'default' and server 'None'
Restoring: default-ed296d619dd6-2017-05-02-220642.psql
Restore tempfile created: 1018.5 KiB
Are you sure you want to continue? [Y/n] y
Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/__init__.py", line 363, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/__init__.py", line 355, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/management/commands/dbrestore.py", line 54, in handle
    self._restore_backup()
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/management/commands/dbrestore.py", line 94, in _restore_backup
    self.connector.restore_dump(input_file)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/base.py", line 96, in restore_dump
    result = self._restore_dump(dump)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/postgresql.py", line 78, in _restore_dump
    return super(PgDumpGisConnector, self)._restore_dump(dump)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/postgresql.py", line 53, in _restore_dump
    stdout, stderr = self.run_command(cmd, stdin=dump, env=self.restore_env)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/postgresql.py", line 20, in run_command
    return super(PgDumpConnector, self).run_command(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/base.py", line 150, in run_command
    "Error running: {}\n{}".format(command, stderr.read()))
dbbackup.db.exceptions.CommandConnectorError: Error running:  psql postgres --host=proj_postgis --port=5432 --username=postgres --no-password --set ON_ERROR_STOP=on 
b'ERROR:  schema "public" does not exist\n'



root@644a3e82f4f5:/code# python3 manage.py dbshell --settings=proj.settings.local
psql (9.5.6, server 9.5.5)
Type "help" for help.

postgres=# create schema public;
CREATE SCHEMA
postgres=# \q
root@644a3e82f4f5:/code# python3 manage.py dbrestore --settings=proj.settings.local
Finding latest backup
Restoring backup for database 'default' and server 'None'
Restoring: default-ed296d619dd6-2017-05-02-220642.psql
Restore tempfile created: 1018.6 KiB
Are you sure you want to continue? [Y/n] y
Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/__init__.py", line 363, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/__init__.py", line 355, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.5/dist-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/management/commands/dbrestore.py", line 54, in handle
    self._restore_backup()
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/management/commands/dbrestore.py", line 94, in _restore_backup
    self.connector.restore_dump(input_file)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/base.py", line 96, in restore_dump
    result = self._restore_dump(dump)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/postgresql.py", line 78, in _restore_dump
    return super(PgDumpGisConnector, self)._restore_dump(dump)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/postgresql.py", line 53, in _restore_dump
    stdout, stderr = self.run_command(cmd, stdin=dump, env=self.restore_env)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/postgresql.py", line 20, in run_command
    return super(PgDumpConnector, self).run_command(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/dbbackup/db/base.py", line 150, in run_command
    "Error running: {}\n{}".format(command, stderr.read()))
dbbackup.db.exceptions.CommandConnectorError: Error running:  psql postgres --host=proj_postgis --port=5432 --username=postgres --no-password --set ON_ERROR_STOP=on 
b'ERROR:  relation "public.users_userprofile" does not exist\n'

```
Hope that helps for a bugfix.

**Comment by psychok7** [comment]
This new issue that i am having might also be related to https://github.com/django-dbbackup/django-dbbackup/issues/230

**Comment by ZuluPro** [comment]
@psychok7 : Really sorry to read that,
Let me some time, I already began Postgres Binary Connector.
I'll add it and check Django 1.11.

**Comment by psychok7** [comment]
@ZuluPro that is really good news. let me know if you need help testing it when your done.

Thanks

**Comment by ZuluPro** [comment]
@psychok7 See PR #241

**Comment by psychok7** [comment]
This PR fixes these problems.
